src/test/libraryfilesynctest.cpp

```diff
@@ -0,0 +1,712 @@
+    }
+
+    TrackPointer prepareTestTrack() const override {
+        TrackPointer pTrack;
+        loadTrackFromDatabaseAndVerifySourceSyncStatus(
+                &pTrack,
+                mixxx::TrackRecord::SourceSyncStatus::Synchronized);
+        return pTrack;
+    }
+
+    void checkTrackRecordSourceSyncStatus() const {
+        LibraryFileSyncTest::checkTrackRecordSourceSyncStatus(
+                mixxx::TrackRecord::SourceSyncStatus::Synchronized);
+    }
+};
+
+class LibraryFileWithMetadataSyncStatusSynchronizedTest
+        : public LibraryFileSyncStatusSynchronizedTest {
+  public:
+    LibraryFileWithMetadataSyncStatusSynchronizedTest()
+            : LibraryFileSyncStatusSynchronizedTest(kTestFileWithMetadata) {
+    }
+};
+
+TEST_F(LibraryFileWithMetadataSyncStatusSynchronizedTest, checkTrackRecordSourceSyncStatus) {
+    LibraryFileSyncStatusSynchronizedTest::checkTrackRecordSourceSyncStatus();
+}
+
+TEST_F(LibraryFileWithMetadataSyncStatusSynchronizedTest, saveTrackMetadataWithSyncEnabled) {
+    modifyAndSaveTrack(prepareTestTrack(), FileType::WithMetadata, true);
+}
+
+TEST_F(LibraryFileWithMetadataSyncStatusSynchronizedTest, saveTrackMetadataWithSyncDisabled) {
+    modifyAndSaveTrack(prepareTestTrack(), FileType::WithMetadata, false);
+}
+
+class LibraryFileWithoutMetadataSyncStatusSynchronizedTest
+        : public LibraryFileSyncStatusSynchronizedTest {
+  public:
+    LibraryFileWithoutMetadataSyncStatusSynchronizedTest()
+            : LibraryFileSyncStatusSynchronizedTest(kTestFileWithoutMetadata) {
+    }
+};
+
+TEST_F(LibraryFileWithoutMetadataSyncStatusSynchronizedTest, checkTrackRecordSourceSyncStatus) {
+    LibraryFileSyncStatusSynchronizedTest::checkTrackRecordSourceSyncStatus();
+}
+
+TEST_F(LibraryFileWithoutMetadataSyncStatusSynchronizedTest, saveTrackMetadataWithSyncEnabled) {
+    modifyAndSaveTrack(prepareTestTrack(), FileType::WithoutMetadata, true);
+}
+
+TEST_F(LibraryFileWithoutMetadataSyncStatusSynchronizedTest, saveTrackMetadataWithSyncDisabled) {
+    modifyAndSaveTrack(prepareTestTrack(), FileType::WithoutMetadata, false);
+}
+
+class LibraryFileSyncStatusOutdatedTest : public LibraryFileSyncTest {
+  public:
+    explicit LibraryFileSyncStatusOutdatedTest(const QFileInfo& testFile)
+            : LibraryFileSyncTest(testFile) {
+    }
+
+  protected:
+    TrackPointer prepareTestTrack() const override {
+        TrackPointer pTrack;
+        establishSourceSyncStatus(&pTrack);
+        return pTrack;
+    }
+
+    void checkTrackRecordSourceSyncStatus() const {
+        LibraryFileSyncTest::checkTrackRecordSourceSyncStatus(
+                mixxx::TrackRecord::SourceSyncStatus::Outdated);
+    }
+
+  private:
+    void establishSourceSyncStatus(TrackPointer* pLoadedTrack) const {
+        // Touch the file's modification time stamp to simulate an external
+        // modification by a 3rd party app after the track has been loaded.
+        updateFileLastModified();
```

Comment:
I don't think so. Let's leave it as is and you may try later if you have the time to do so.

Reply:
We cannot reorder the code because the database includes the check of the desired status that would fail if we modify the time stamp later.

Reply:
updateFileLastModifiedAndSleep(): This naming would be wrong, because it reveals an implementation detail that should be hidden from the caller.

Once again: The invocation of sleep does not belong to the test routine! It is only required for technical reasons because we have no control of the side effects cause by the file system access.

Reply:
A function should be named like what it does and not hide important information. But I prefer anyway to move all the sleeps one level above. That would make the code much easier to read and the test more consistent.